### ISSUE TITLE
Document failure behavior on declarative client

### ISSUE THREAD

**Issue by chrylis** [micronaut-projects/micronaut-core#1188]
The current documentation does not explain the behavior of the declarative client in error conditions; for example, there is the `HttpClientException` hierarchy, but it is not clear under what circumstances which exceptions, if any, are thrown; for example, a blocking request returning a POJO seems to be actually returning `null` on a 404 when `HttpClientResponseException` is expected:

```
2019-01-29 17:32:13.891 DEBUG 30214 --- [ntLoopGroup-1-4] i.m.http.client.DefaultHttpClient        : Sending HTTP Request: GET /users/c1f34d23-24f3-406a-a9f1-031a3c37311d/profile/contact/mobile
2019-01-29 17:32:13.891 DEBUG 30214 --- [ntLoopGroup-1-4] i.m.http.client.DefaultHttpClient        : Chosen Server: localhost(8091)
2019-01-29 17:32:13.891 TRACE 30214 --- [ntLoopGroup-1-4] i.m.http.client.DefaultHttpClient        : Accept: application/json
2019-01-29 17:32:13.891 TRACE 30214 --- [ntLoopGroup-1-4] i.m.http.client.DefaultHttpClient        : host: localhost:8091
2019-01-29 17:32:13.891 TRACE 30214 --- [ntLoopGroup-1-4] i.m.http.client.DefaultHttpClient        : connection: close
2019-01-29 17:32:14.100 TRACE 30214 --- [ntLoopGroup-1-4] i.m.http.client.DefaultHttpClient        : HTTP Client Response Received for Request: GET http://localhost:8091/users/c1f34d23-24f3-406a-a9f1-031a3c37311d/profile/contact/mobile
2019-01-29 17:32:14.100 TRACE 30214 --- [ntLoopGroup-1-4] i.m.http.client.DefaultHttpClient        : Status Code: 404 
2019-01-29 17:32:14.100 TRACE 30214 --- [ntLoopGroup-1-4] i.m.http.client.DefaultHttpClient        : Content-Type: application/json;charset=UTF-8
2019-01-29 17:32:14.100 TRACE 30214 --- [ntLoopGroup-1-4] i.m.http.client.DefaultHttpClient        : Date: Wed, 30 Jan 2019 01:32:14 GMT
2019-01-29 17:32:14.100 TRACE 30214 --- [ntLoopGroup-1-4] i.m.http.client.DefaultHttpClient        : Connection: close
2019-01-29 17:32:14.100 TRACE 30214 --- [ntLoopGroup-1-4] i.m.http.client.DefaultHttpClient        : content-length: 243
2019-01-29 17:32:14.101 TRACE 30214 --- [ntLoopGroup-1-4] i.m.http.client.DefaultHttpClient        : Response Body
2019-01-29 17:32:14.101 TRACE 30214 --- [ntLoopGroup-1-4] i.m.http.client.DefaultHttpClient        : ----
2019-01-29 17:32:14.101 TRACE 30214 --- [ntLoopGroup-1-4] i.m.http.client.DefaultHttpClient        : {"timestamp":1548811934090,"status":404,"error":"Not Found","exception":"userprofile.web.NotFoundException","message":"No message available","path":"/users/c1f34d23-24f3-406a-a9f1-031a3c37311d/profile/contact/mobile"}
2019-01-29 17:32:14.101 TRACE 30214 --- [ntLoopGroup-1-4] i.m.http.client.DefaultHttpClient        : ----
2019-01-29 17:32:14.166 ERROR 30214 --- [nio-8090-exec-1] o.a.c.c.C.[.[.[/].[dispatcherServlet]    : Servlet.service() for servlet [dispatcherServlet] in context with path [] threw exception [Request processing failed; nested exception is java.lang.NullPointerException] with root cause

java.lang.NullPointerException: null
	at Sender.sendConfirmationCode(Sender.java:80) ~[classes/:na]
```

**Comment by FylmTM** [comment]
Hi!

Just stumbled upon this problem with `404`.
Was writing test using Micronaut client, for an endpoint that return 404 when record is not found in database.
Received `null` as a response, instead of `HttpClientResponseException` that I expected.

Accordingly to `io/micronaut/http/client/DefaultHttpClient.java:1742` all >= 400 status codes should be an exception.
But later I found that `NOT_FOUND` is specifically treated differently.

Questions:
- I can't find anywhere a reason on why `NOT_FOUND` is handled differently. Is there one?
- Is there any way how to force Micronaut client to throw exceptions for all 4** statuses, including `NOT_FOUND`?

Thanks.

**Comment by graemerocher** [comment]
The behaviour of Micronaut's declarative client mirrors other popular declarative clients like OpenFeign and treats a 404 as a non exceptional condition. 

The reason is that if it were an exception it would trigger retry/failover/fallback logic which makes no sense for 404s. The explanation for the behaviour can be found in this commit in Feign:

https://github.com/OpenFeign/feign/commit/24885fe9620ed620af43f4d2d6ffcfc82980e097

Feel free to send a PR to the documentation if you wish to see improvements.

**Comment by chrylis** [comment]
Unfortunately, this is a chicken-and-egg first issue--I don't know what the design-spec behavior is supposed to be (and the nature of Micronaut makes figuring that out from reading the code nontrivial).

**Comment by TWiStErRob** [comment]
Is there a way to turn this off with config? When integrating an API one might get a 404 because the URL is wrong (typo). I had to turn on trace logging and research for hours trying to figure out why I'm getting a `null`. Once I saw the message in the trace log, it was trivial to fix. I want my code to FAIL HARD when there's a problem.

Few hours later again, I discovered that [GitHub will also send a 404 when auth is missing...](https://docs.github.com/en/rest/using-the-rest-api/troubleshooting-the-rest-api?apiVersion=2022-11-28#404-not-found-for-an-existing-resource). i.e. "your token is wrong" -> `null`.

Out of curiosity, where is this implemented in declarative micronaut HTTPClient? I couldn't find it yet.
